packages/@react-spectrum/menu/src/useSafelyMouseToSubmenu.ts

```diff
@@ -0,0 +1,118 @@
+import {MutableRefObject, RefObject, useEffect, useRef} from 'react';
+
+interface SafeMouseToSubmenuOptions {
+  /** Ref for the parent menu. */
+  menuRef: RefObject<HTMLElement>,
+  /** Ref for the submenu. */
+  submenuRef: MutableRefObject<any>,
+  /** Ref for the submenu's trigger element. */
+  triggerRef: RefObject<HTMLElement>,
+  /** Whether the submenu is open. */
+  isOpen: boolean
+}
+
+/**
+ * Allows the user to move their pointer to the submenu without it closing when their mouse leaves the trigger element.
+ * Prevents pointer events from going to the underlying menu if the user is moving their pointer towards the sub-menu.
+ */
+export function useSafelyMouseToSubmenu(options: SafeMouseToSubmenuOptions) {
+  let {menuRef, submenuRef, triggerRef, isOpen} = options;
+  let prevPointerPos = useRef<{x: number, y: number, time: number}>(null);
+  let currentSpeed = useRef<number>(null);
+  let menuRect = useRef<DOMRect>(null);
+  let submenuRect = useRef<DOMRect>(null);
+  let triggerRect = useRef<DOMRect>(null);
+  let lastProcessedTime = useRef<number>(0);
+
+  useEffect(() => {
+    let menu = menuRef.current;
+    let submenu = (submenuRef.current.UNSAFE_getDOMNode() as HTMLElement);
+
+    if (!menu || !submenu || !isOpen) {
+      return;
+    }
+
+    menuRect.current = menu.getBoundingClientRect();
+    submenuRect.current = submenu.getBoundingClientRect();
+    triggerRect.current = triggerRef.current.getBoundingClientRect();
+
+    let updateMenuRects = () => {
+      menuRect.current = menu.getBoundingClientRect();
+      submenuRect.current = submenu.getBoundingClientRect();
+    };
+
+    window.addEventListener('resize', updateMenuRects);
```

Comment:
i think we should attach a resize observer to the menu/trigger/submenu, otherwise we won't update for async lists